### PR TITLE
Add support for environment variable substitution

### LINKED ISSUE #224: version numbers / image tags in kedge YAML
whats the current approach for replacing version numbers in a kedge YAML? Am thinking of CI / CD pipelines that tend to generate new unique release numbers for each new version and tend to put those versions in the YAML as a version label and then use the docker image tag of the same name as the version

should we use environment variable expressions or something? e.g.

```yaml
name: foo
labels:
  version: "${VERSION}"
image: "foo/bar:${VERSION}"
```
then we can specify it via something like...
```
export VERSION="1.2.3"
kedge create -f foo.yaml (or whatever the command is ;)
```

### PR DESCRIPTION
**Pull request by kadel**
You can use environment variables in you Kedge files. Expression is `[[ variable_name ]]`.

fixes #224 

TODO:
- ~~move variable substitution logic from pkg/cmd/util.go to its own package~~
- [x] documentation

> ~~move variable substitution logic from pkg/cmd/util.go to its own package~~

I'm keeping this as it is, for now. We need clean up whole `pkg/cmd` package. Created https://github.com/kedgeproject/kedge/issues/301

### CODE REVIEW

File: pkg/cmd/util.go [hunk 1]
```diff
@@ -107,3 +107,49 @@ func GetAllYAMLFiles(paths []string) ([]string, error) {
...
+
+	if len(remainingVariables) > 0 {
+		return nil, fmt.Errorf("undefined variable(s): %s", strings.Join(remainingVariables, ", "))
+	}
+
+	return newContent, nil
+}
+
+// replaceWithEnv is used with regexp.ReplaceAllFunc to replace variable
+// with value from envrironment
```
Comment (containscafeine) [inline pkg/cmd/util.go]: nit: environment

Comment (kadel) [reply pkg/cmd/util.go]: typo fixed, thx

**Comment by surajssd** [comment]
@kadel @containscafeine 

Am I doing anything wrong?

```console
$ kedge version
0.2.0 (f9c6b37)

$ cat file.yaml 
name: httpd
containers:
- image: [[ imagename ]]

$ imagename=foo
$ echo $imagename
foo

$ kedge generate -f .
failed to replace variables: undefined variable(s): imagename
```

**Comment by surajssd** [comment]
This worked alright:

```console
$ export imagename=foo
$ echo $imagename
foo

$ kedge generate -f .
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: httpd
  name: httpd
spec:
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: httpd
      name: httpd
    spec:
      containers:
      - image: foo
        name: httpd
        resources: {}
status: {}
```

**Comment by surajssd** [comment]
```console
$ imagename=foo kedge generate -f .
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: httpd
  name: httpd
spec:
...
```

This also worked fine!

**Comment by kadel** [comment]
> @kadel @containscafeine
> 
> Am I doing anything wrong?
> 
> $ kedge version
> 0.2.0 (f9c6b37)
> 
> $ cat file.yaml 
> name: httpd
> containers:
> - image: [[ imagename ]]
> 
> $ imagename=foo
> $ echo $imagename
> foo
> 
> $ kedge generate -f .
> failed to replace variables: undefined variable(s): imagename

This is OK it shouldn't work. This is how env variables work in bash.

If you don't export it the variable is not passed to any child processes of given shell.




```
imagename=foo ./kedge
```
Above works because In this case variable is directly passed to kedge process.



```
export imagename=foo
 ./kedge
```
Above is also working because export is used and that means that any child process of given shell will also see this variable


But following doesn't work as kedge process is a new child process of given shell, and it doesn't receive unexported variables.
```
imagename=foo 
./kedge
```


it works with echo, because echo is build in bash function.

**Pull request merged**